core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,296 @@
+
+    private static final ObjectMapper mapper = JacksonUtil.newDefaultObjectMapper();
+
+    /**
+     * Generate an array of json schema specifications for each method inside the service.
+     * @param serviceSpecification the service specification to generate the json schema from.
+     * @return ArrayNode that contains service specifications
+     */
+    public static ArrayNode generate(ServiceSpecification serviceSpecification) {
+        // TODO: Test for Thrift, GraphQL, and annotated services
+        final JsonSchemaGenerator generator = new JsonSchemaGenerator(serviceSpecification);
+        return generator.generate();
+    }
+
+    private final Set<EnumInfo> enumInfos;
+    private final Set<StructInfo> structInfos;
+    private final Set<ServiceInfo> serviceInfos;
+    private final Map<String, StructInfo> typeNameToStructMapping;
+    private final Map<String, EnumInfo> typeNameToEnumMapping;
+
+    private JsonSchemaGenerator(ServiceSpecification serviceSpecification) {
+        this.enumInfos = serviceSpecification.enums();
+        this.structInfos = serviceSpecification.structs();
+        this.serviceInfos = serviceSpecification.services();
+        this.typeNameToStructMapping = structInfos.stream().collect(
+                Collectors.toMap(StructInfo::name, Function.identity()));
+        this.typeNameToEnumMapping = enumInfos.stream().collect(
+                Collectors.toMap(EnumInfo::name, Function.identity()));
+    }
+
+    private ArrayNode generate() {
+        final ArrayNode definitions = mapper.createArrayNode();
+
+        final Set<ObjectNode> methodDefinitions = serviceInfos.stream().flatMap(
+                serviceInfo -> serviceInfo.methods().stream()
+                                          .map(methodInfo -> generate(serviceInfo, methodInfo))).collect(
+                Collectors.toSet());
+
+        return definitions.addAll(methodDefinitions);
+    }
+
+    /**
+     * Generate the JSON Schema for the given {@link ServiceInfo} and {@link MethodInfo}.
+     * @param serviceInfo service info.
+     * @param methodInfo the method to generate the JSON Schema for.
+     * @return ObjectNode containing the JSON schema for the parameter type.
+     */
+    private ObjectNode generate(ServiceInfo serviceInfo, MethodInfo methodInfo) {
+        final ObjectNode root = mapper.createObjectNode();
+
+        root.put("$id", methodInfo.id())
+            .put("title", methodInfo.name())
+            .put("description", methodInfo.descriptionInfo().docString())
+            .put("additionalProperties", false)
+            // TODO: Assumes every method takes an object, which is only valid for RPC based services
+            //  and most of the REST services.
+            .put("type", "object");
+
+        // Workaround for gRPC services because we should unwrap the first "request" parameter.
+        final boolean isProto = methodInfo.endpoints().stream().flatMap(x -> x.availableMimeTypes().stream())
```

Comment:
How about adding a boolean parameter when creating a `MethodInfo` and retrieving it via `MethodInfo.isGrpc()`?

Note for reviewers:
The layout of JSON requests for gRPC and Thrift are different. gRPC uses the method parameter as root but Thrift uses the method parameters with the key. For example:
```
// When struct is like:
MyStruct {
  String name;
  int age;
}

// When the method is `myMethod(MyStruct myStruct)`
// for gRPC:
{
  "name": "foo",
  "age": 40
}

// for Thrift:
{
  "myStruct": {
    "name": "foo",
    "age": 40
  }
}
```

Reply:
> How about adding a boolean parameter when creating a MethodInfo and retrieving it via MethodInfo.isGrpc()?

Good idea but it will require lots of file changes (even if they are small), how about opening an issue and handling it separately?

Reply:
That's a good idea. 😄 Let me handle that.

Reply:
https://github.com/line/armeria/pull/4592
Created. 😄